cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+
+3. Challenge: The leader computes a collective challenge
+c = H( Aggregate Commit ∥ Aggregate Public key || Message ),
+then multicasts c down through the tree, along
+with the statement S to be signed if it was not already
+announced in phase 1.
+
+4. Response: In a final bottom-up phase, each node i waits
+to receive a partial aggregate response rˆj from each of
+its immediate children j ∈ Ci. Node i now computes its
+individual response ri = vi + cxi, and its partial aggregate
+response rˆi = ri + \sum{j ∈ Cj}{rˆj} . Node i finally passes rˆi
+up to its parent, unless i is the root.
+*/
+package cosi
+
+import (
+	cryptoRand "crypto/rand"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/dedis/crypto/abstract"
+	//own "github.com/nikkolasg/learning/crypto/util"
+)
+
+// Cosi is the struct that implements the basic cosi.
+type Cosi struct {
+	// Suite used
+	suite abstract.Suite
+	// publics is the list of public keys used for signing
+	publics []abstract.Point
+	// mask is the mask used to select which signers participated in this round
+	// or not. All code regarding the mask is directly inspired from
+	// github.com/bford/golang-x-crypto/ed25519/cosi code.
+	mask *CosiMask
+	// the longterm private key we use during the rounds
+	private abstract.Secret
+	// the message being co-signed
+	message []byte
+	// timestamp of when the announcement is done (i.e. timestamp of the four
+	// phases)
+	timestamp int64
+	// random is our own secret that we wish to commit during the commitment phase.
+	random abstract.Secret
+	// commitment is our own commitment
+	commitment abstract.Point
+	// V_hat is the aggregated commit (our own + the children's)
+	aggregateCommitment abstract.Point
+	// globalComitment is the global commit of all participants
+	//passed down in the challenge phase
+	globalCommitment abstract.Point
+	// challenge holds the challenge for this round
+	challenge abstract.Secret
+	// response is our own computed response
+	response abstract.Secret
+	// aggregateResponses is the aggregated response from the children + our own
+	aggregateResponse abstract.Secret
+}
+
+// NewCosi returns a new Cosi struct given the suite, the longterm secret, and
+// the list of public keys. If some signers were not to be participating, you
+// have to set the mask using `SetMask` method. By default, all participants are
+// designated as participating.
+func NewCosi(suite abstract.Suite, private abstract.Secret, publics []abstract.Point) *Cosi {
+	cos := &Cosi{
+		suite:   suite,
+		private: private,
+		publics: publics,
+	}
+	// Start with an all-disabled participation mask, then set it correctly
+	cos.mask = NewCosiMask(suite, publics)
+	return cos
+}
+
+// Announcement holds only the timestamp for that round
+type Announcement struct {
+	Timestamp int64
+}
+
+// Commitment sends it's own commit Vi and the aggregate children's
+// commit V^i
+type Commitment struct {
+	Commitment     abstract.Point
+	ChildrenCommit abstract.Point
+}
```

Comment:
It's not really about failure but about mischief, if a node does not generate correctly its response, we can check only its individual response (by having its individual commit). Instead of checking for the whole subtree if the response is correct, we can check also for direct children's reponse too. But if you think that's not a good use case, I'm all for removing it also, it's not so important because if a wrong response has been generated, we need to deal with it anyway by restarting I guess so..